frame/support/tests/versioned_runtime_upgrade.rs

```diff
@@ -0,0 +1,259 @@
+
+use frame_support::{
+	construct_runtime, derive_impl,
+	migrations::VersionedRuntimeUpgrade,
+	parameter_types,
+	traits::{ConstU32, GetStorageVersion, OnRuntimeUpgrade, StorageVersion},
+	weights::constants::RocksDbWeight,
+};
+use frame_system::Config;
+use sp_core::{ConstU16, Get};
+
+#[cfg(feature = "try-runtime")]
+use once_cell::sync::Lazy;
+#[cfg(feature = "try-runtime")]
+use std::sync::Mutex;
+
+type UncheckedExtrinsic = frame_system::mocking::MockUncheckedExtrinsic<Test>;
+type Block = frame_system::mocking::MockBlock<Test>;
+
+#[frame_support::pallet]
+mod dummy_pallet {
+	use frame_support::pallet_prelude::*;
+
+	const STORAGE_VERSION: StorageVersion = StorageVersion::new(4);
+
+	#[pallet::pallet]
+	#[pallet::storage_version(STORAGE_VERSION)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {}
+
+	#[pallet::storage]
+	pub type SomeStorage<T: Config> = StorageValue<_, u32, ValueQuery>;
+
+	#[pallet::genesis_config]
+	#[derive(Default)]
+	pub struct GenesisConfig {}
+
+	#[pallet::genesis_build]
+	impl<T: Config> GenesisBuild<T> for GenesisConfig {
+		fn build(&self) {}
+	}
+}
+
+impl dummy_pallet::Config for Test {}
+
+construct_runtime!(
+	pub enum Test where
+		Block = Block,
+		NodeBlock = Block,
+		UncheckedExtrinsic = UncheckedExtrinsic,
+	{
+		System: frame_system::{Pallet, Call, Config, Storage, Event<T>} = 0,
+		DummyPallet: dummy_pallet::{Pallet, Config, Storage} = 1,
+	}
+);
+
+#[derive_impl(frame_system::config_preludes::TestDefaultConfig as frame_system::DefaultConfig)]
+impl frame_system::Config for Test {
+	type BaseCallFilter = frame_support::traits::Everything;
+	type RuntimeOrigin = RuntimeOrigin;
+	type RuntimeCall = RuntimeCall;
+	type RuntimeEvent = RuntimeEvent;
+	type PalletInfo = PalletInfo;
+	type OnSetCode = ();
+}
+
+pub(crate) fn new_test_ext() -> sp_io::TestExternalities {
+	let storage = frame_system::GenesisConfig::default().build_storage::<Test>().unwrap();
+	let mut ext: sp_io::TestExternalities = sp_io::TestExternalities::from(storage);
+	ext.execute_with(|| System::set_block_number(1));
+	ext
+}
+
+/// A dummy migration for testing the `VersionedRuntimeUpgrade` trait.
+/// Sets SomeStorage to S.
+struct SomeUnversionedMigration<T: Config, S: Get<u32>>(sp_std::marker::PhantomData<(T, S)>);
+
+parameter_types! {
+	const UpgradeReads: u64 = 4;
+	const UpgradeWrites: u64 = 2;
+}
+
+#[cfg(feature = "try-runtime")]
+static PRE_UPGRADE_RETURN_BYTES: [u8; 4] = [0, 1, 2, 3];
+
+// We can't write to pallet storage from pre/post hooks, so use a global variable to track that they
+// are called correctly.
+#[cfg(feature = "try-runtime")]
+static PRE_UPGRADE_CALLED: Lazy<Mutex<bool>> = Lazy::new(|| Mutex::new(false));
+#[cfg(feature = "try-runtime")]
+static POST_UPGRADE_CALLED: Lazy<Mutex<bool>> = Lazy::new(|| Mutex::new(false));
+#[cfg(feature = "try-runtime")]
+static POST_UPGRADE_CALLED_WITH: Lazy<Mutex<Vec<u8>>> = Lazy::new(|| Mutex::new(Vec::new()));
```

Comment:
This is pretty messy. Any suggestions for not needing to repeat `#[cfg(feature = "try-runtime")]`?

Reply:
You can put it into a `mod`, or just do `allow_unsed` or whatever it reports.